Mathlib/Analysis/PSeries.lean

```diff
@@ -17,132 +17,218 @@ converges if and only if so does `∑ k, 2 ^ k f (2 ^ k)`. We prove this test in
-#align ennreal.tsum_condensed_le ENNReal.tsum_condensed_le
-
+          (mul_le_mul_of_nonneg_left (ENNReal.sum_le_tsum <| Finset.Ico (u 0 + 1) (u n + 1)) _) _)
+  apply hC
+  replace hCpos : C > (0 : ℝ≥0∞) := ENNReal.coe_lt_coe.2 hCpos
+  apply le_of_lt hCpos
 end ENNReal
 
 namespace NNReal
-
-/-- Cauchy condensation test for a series of `NNReal` version. -/
-theorem summable_condensed_iff {f : ℕ → ℝ≥0} (hf : ∀ ⦃m n⦄, 0 < m → m ≤ n → f n ≤ f m) :
-    (Summable fun k : ℕ => (2 : ℝ≥0) ^ k * f (2 ^ k)) ↔ Summable f := by
-  simp only [← ENNReal.tsum_coe_ne_top_iff_summable, Ne.def, not_iff_not, ENNReal.coe_mul,
-    ENNReal.coe_pow, ENNReal.coe_two]
+/-- for a series of `NNReal` version. -/
+theorem summable_schlomilch_iff {u : ℕ → ℕ} {f : ℕ → ℝ≥0} (hf : ∀ ⦃m n⦄, 0 < m → m ≤ n → f n ≤ f m)
+    (h_pos : ∀ n, 0 < u n) (strict_mono : ∀ ⦃m n⦄, 0 ≤ m → m < n → u m < u n)
+    (h_succ_diff : succ_diff_bounded u) :
+    (Summable fun k : ℕ => (u (k + 1) - (u k : ℝ≥0)) * f (u k)) ↔ Summable f := by
+  simp only [← ENNReal.tsum_coe_ne_top_iff_summable, Ne.def, not_iff_not, ENNReal.coe_mul]
   constructor <;> intro h
   · replace hf : ∀ m n, 1 < m → m ≤ n → (f n : ℝ≥0∞) ≤ f m := fun m n hm hmn =>
       ENNReal.coe_le_coe.2 (hf (zero_lt_one.trans hm) hmn)
-    simpa [h, ENNReal.add_eq_top, ENNReal.mul_eq_top] using ENNReal.tsum_condensed_le hf
+    obtain ⟨C, hCpos, hC⟩ := ENNReal.tsum_schlomilch_le hf h_pos strict_mono h_succ_diff
+    have C_nonzero : C ≠ 0 := ne_of_gt hCpos
+    have : (↑(u 1) - ↑(u 0)) * ↑(f (u 0)) + ↑C * ∑' (k : ℕ), ↑(f k) = ∞ := by exact eq_top_mono hC h
+    simpa [ENNReal.add_eq_top, ENNReal.mul_ne_top, ENNReal.mul_eq_top, C_nonzero]
   · replace hf : ∀ m n, 0 < m → m ≤ n → (f n : ℝ≥0∞) ≤ f m := fun m n hm hmn =>
       ENNReal.coe_le_coe.2 (hf hm hmn)
-    simpa [h, ENNReal.add_eq_top] using ENNReal.le_tsum_condensed hf
-#align nnreal.summable_condensed_iff NNReal.summable_condensed_iff
-
+    have : ∑ k in range (u 0), ↑(f k) ≠ ∞ := by
+      apply ne_top_of_lt (ENNReal.sum_lt_top fun a _ => coe_ne_top)
+    simpa [h, ENNReal.add_eq_top, this] using ENNReal.le_tsum_schlomilch hf h_pos strict_mono
 end NNReal
 
-/-- Cauchy condensation test for series of nonnegative real numbers. -/
+/-- for series of nonnegative real numbers. -/
+theorem summable_schlomilch_iff_of_nonneg {u : ℕ → ℕ} {f : ℕ → ℝ} (h_nonneg : ∀ n, 0 ≤ f n)
+    (hf : ∀ ⦃m n⦄, 0 < m → m ≤ n → f n ≤ f m) (h_pos : ∀ n, 0 < u n)
+    (strict_mono : ∀ ⦃m n⦄, 0 ≤ m → m < n → u m < u n) (h_succ_diff : succ_diff_bounded u) :
+    (Summable fun k : ℕ => (u (k + 1) - (u k : ℝ)) * f (u k)) ↔ Summable f := by
+  lift f to ℕ → ℝ≥0 using h_nonneg
+  simp only [NNReal.coe_le_coe] at *
+  have : ∀ k, (u (k + 1) - (u k : ℝ)) = ((u (k + 1) : ℝ≥0) - (u k : ℝ≥0) : ℝ≥0) := by
+    intro k
+    rw [@coe_sub_def]
+    rw [max_eq_left]
+    simp only [NNReal.coe_nat_cast]
+    refine sub_nonneg_of_le ?_
+    norm_cast
+    exact (strict_mono k.zero_le k.lt_succ_self).le
```

Comment:
a short golf. Note also how you can pass parameters to `have` just like declarations (i.e., the `(k : ℕ)`) so you don't have to write `intro k`.
```suggestion
  have (k : ℕ) : (u (k + 1) - (u k : ℝ)) = ((u (k + 1) : ℝ≥0) - (u k : ℝ≥0) : ℝ≥0) := by
    have := Nat.cast_le (α := ℝ≥0).mpr <| (strict_mono (zero_le k) k.lt_succ_self).le
    simp [NNReal.coe_sub this]
```